core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1254,7 +1265,10 @@ void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, ExecutionE
                         canProcess = false;
                         continue;
                     }
-                    if (RemoteLogSegmentState.DELETE_SEGMENT_FINISHED.equals(metadata.state())) {
+                    // skip the COPY_SEGMENT_STARTED segments since they might be the dangling segments that failed before
+                    // and blocks the normal segment deletion, ex: it failed `isRemoteSegmentWithinLeaderEpochs` check... etc
+                    if (RemoteLogSegmentState.DELETE_SEGMENT_FINISHED.equals(metadata.state()) ||
+                            RemoteLogSegmentState.COPY_SEGMENT_STARTED.equals(metadata.state())) {
```

Comment:
Can we avoid this check?  The pending dangling segments (that was missed to delete in the copy-phase due to remote storage non-availability during that time) gets removed when they breach the retention time / size.

> blocks the normal segment deletion

The normal segment gets deleted in the next run since we calculate the size based on the COPY_SEGMENT_FINISHED and DELETE_SEGMENT_STARTED states.

Reply:
Fine to me. I just wanted to skip  these dangling segments to avoid to cause any unexpected issues. But I can't think of any issues so far.

Reply:
Ok, got the reason for including the `COPY_SEGMENT_STARTED` state together with this check. When the dangling segments gets deleted, the event state is still set to "COPY_SEGMENT_STARTED" and the expiry thread will call the RemoteStorageManager#deleteLogSegmentData again. For my understanding, why are we not changing the state to `DELETE_SEGMENT_FINISHED`?

I think it should be ok, since we mentioned the below in the javadoc contract:

https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java?L146
```
     * This operation is expected to be idempotent. If resources are not found, it is not expected to
     * throw {@link RemoteResourceNotFoundException} as it may be already removed from a previous attempt
```

Reply:
> For my understanding, why are we not changing the state from COPY_SEGMENT_STARTED -> DELETE_SEGMENT_STARTED -> DELETE_SEGMENT_FINISHED?

Yes, why not? Created [KAFKA-17428](https://issues.apache.org/jira/browse/KAFKA-17428) for this issue.

Reply:
I thought about this when I first had a look at it and one thing that bothered me is that `DELETE_SEGMENT_STARTED` means to me that we're now in a state where we attempt deletion. However if the remote store is down and we fail to copy and delete we will leave that segment in `DELETE_SEGMENT_STARTED` and not attempt to delete it till the segment itself breaches retention.ms/bytes. 

We can probably just make it clearer but that was my thought at the time.

Reply:
Yes, that's indeed a place we can improve! I've added the comment into [KAFKA-17428](https://issues.apache.org/jira/browse/KAFKA-17428). Thanks.